Turn dataset_query into structured data in search results

dpsutton opened the pull request:
Fixes #15959 

Search results query a card's dataset query as text but we need to send it back as structured data afterwards

#### Before
```javascript
{
      "name": "bar - races count",
      ...
      "dataset_query": "{\"type\":\"query\",\"query\":{\"source-query\":{\"source-table\":62,\"aggregation\":[[\"count\"]],\"breakout\":[[\"field\",379,null]]},\"filter\":[\">\",[\"field\",\"count\",{\"base-type\":\"type/Integer\"}],20]},\"database\":13}",
      "id": 5867,
            
    },
```

#### After
```javascript
 {
      "name": "bar - races count",
      ...
      "dataset_query": {
        "type": "query",
        "query": {
          "source-query": {
            "source-table": 62,
            "aggregation": [
              [
                "count"
              ]
            ],
            "breakout": [
              [
                "field",
                379,
                null
              ]
            ]
          },
          "filter": [
            ">",
            [
              "field",
              "count",
              {
                "base-type": "type/Integer"
              }
            ],
            20
          ]
        },
        "database": 13
      },
      ...
    },
```

dpsutton committed:
Turn dataset_query into structured data in search results

dpsutton committed:
sort scoring test ns

kulyk reviewed (approved):
Couldn't reproduce #15959, searching questions by SQL query text works fine for me too
Thank you!

dpsutton committed:
Fixup tests

we are no longer returning string results. The cards in the test have
empty queries, which the model turns into {} but normalization removes
empty clauses and returns nil. The tests would be too annoying to hook
this up to work properly and this functionality is tested in other
places.

kulyk commented:
@dpsutton could you please enable the cypress repro test for this issue? It's [here](https://github.com/metabase/metabase/blob/ab75c941e57520195d62612a3d397899e06e042f/frontend/test/metabase/scenarios/dashboard/dashboard.cy.spec.js#L410)

dpsutton committed:
Unskip repro for #15959

dpsutton commented:
@kulyk thanks i always forget about those

kulyk committed:
Update repro test

kulyk commented:
@dpsutton had to update the repro test as it was relying on the old UI for adding questions (a modal instead of a sidebar). Here's it failing on `master`:

<img width="1480" alt="CleanShot 2021-08-31 at 20 19 30@2x" src="https://user-images.githubusercontent.com/17258145/131547963-4284eb6a-edb7-4aa7-a040-8796da451b0d.png">

kulyk commented:
@metabase-bot backport release-x.40.x
